README
NAME
    App::pandoc::preprocess - Preprocess Pandoc before Processing Pandoc

VERSION
    version 0.9.5

ppp - pandoc pre-process
USAGE
         cat chapters/input-*.pandoc | ppp | pandoc -o output.pdf --smart [more pandoc options...]

    Additionally see `etc/input.txt` for concrete examples.

PREREQUISITES
    *   dot/neato (neato is new!)

    *   rdfdot

    *   ditaa

    *   Image::Magick (for downscaling of large images)

BACKGROUND
    *   much simpler design than version 1: pipeable & chainable, reading
        line-by-line

    *   parallelized work on image file creation

  How it works
    1. while-loop will iterate line by line and is using the
    flip-flop-operator:

         * as soon, as a ditaa/rdfdot/dot-block starts,
           globals ($fileno, $outfile, etc) are set, so all other routines can see them
         * when actually *inside* the block, the block's contents are printed
           to the newly generated file (image-X.(ditaa/rdfdot/dot))

    2. once the flip-flop-operator hits the end of the
    ditaa/rdfdot/dot-block, a child will be spawned to take over the actual
    ditaa/rdfdot/dot-process to create the png-file and the globals are
    reset

    3. all other lines which are not part of a ditaa/rdfdot/dot-block will
    simply be piped through to stdout

    4. at the end of the program, all children are waited for

    5. in the meantime, the new pandoc contents are printed to stdout

    6. all child-processes will remain quiert as far as stdout is concerned
    and write to their individual log-files

  Todo
    *   Captions

    *   Checks whether ditaa... are available

    *   check whether ditaa has file.encoding set

    *   bundle ditaa with this

AUTHOR
    DBR <dbr@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by DBR.

    This is free software, licensed under:

      DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE, Version 2, December 2004



Changes
0.9.5     2014-08-11 09:08:13+02:00 Europe/Berlin
  - change directory options to --img and --log 

0.9.4     2014-04-07 17:31:14 Europe/Berlin

  - don't use Zilla::Plugin::Fatpacker: it leaves off the shebang and renders executable useless.
  - better handling of commandline arguments

0.9.0     2014-03-31 23:54:27 Europe/Berlin

  - create tmp/ dir automatically using File::Temp::tempdir

0.7.2     2014-03-13 10:50:49 Europe/Berlin
  
  - update documentation

0.7.1     2014-03-12 23:55:30 Europe/Berlin

  - errors branching: bump release: (finally) add plantuml as renderer

0.7.0     2014-03-12 23:52:00 Europe/Berlin

  - add plantuml as renderer

0.6.6     2014-03-06 14:21:31 Europe/Berlin

  - Update documentation

0.6.5     2014-03-06 00:09:19 Europe/Berlin

  - Assume defaults for yuml unless given

0.6.3     2014-03-05 17:26:09 Europe/Berlin

  - This is stupid: simply rely on yuml's own pythong-package-installation and leave that executable out!

0.6.2     2014-03-05 16:40:10 Europe/Berlin

  - Update README.md

0.6.1     2014-03-05 16:34:05 Europe/Berlin
  
  - Add yUML as renderer

0.5.0     2014-02-07 12:39:07 Europe/Berlin

  - Take attributes into account

0.4.1     2014-01-22 23:29:01 Europe/Berlin

  - add documentation

0.4.0     2014-01-22 15:52:41 Europe/Berlin

  - Refactor/Rewrite: Use flip-flop operator, major simplifications

0.3.1     2014-01-07 17:32:20 Europe/Berlin

  - Pluginbundle should finally work and this module should get indexed

0.3.0     2014-01-07 16:59:41 Europe/Berlin
  
  - Check environment for installed programs (rudimentrily)

0.2.4     2014-01-07 16:50:50 Europe/Berlin

  - "funny characters" in ditaa work

0.2.3     2014-01-07 16:46:26 Europe/Berlin

  - code blocks can appear in lists, too

0.2.2     2014-01-07 16:39:38 Europe/Berlin

  - Add check for ditaa executable

0.2.1     2014-01-06 13:01:48 Europe/Berlin

  - Add moo

0.2.0     2014-01-05 13:00:48 Europe/Berlin

  - Change everything to Moo

0.1.0_1   2014-01-04 11:14:14 Europe/Berlin

  - First alpha release


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.019.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Preprocess Pandoc before Processing Pandoc",
  "dist_author" => [
    "DBR <dbr\@cpan.org>"
  ],
  "dist_name" => "App-pandoc-preprocess",
  "dist_version" => "0.9.5",
  "license" => "unrestricted",
  "module_name" => "App::pandoc::preprocess",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "File::Path" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "perl" => "v5.14.0",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/ppp",
    "bin/ppp"
  ],
  "test_requires" => {
    "Carp" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Scalar::Util" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.94",
    "Test::UseAllModules" => 0
  }
);


my %fallback_build_requires = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.3601",
  "Scalar::Util" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.94",
  "Test::UseAllModules" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/App/pandoc/preprocess.pm
package App::pandoc::preprocess;
BEGIN {
  $App::pandoc::preprocess::AUTHORITY = 'cpan:VOJ';
}
$App::pandoc::preprocess::VERSION = '0.9.5';
#  PODNAME: App::pandoc::preprocess
# ABSTRACT: Preprocess Pandoc before Processing Pandoc


'make CPAN happy -- we only have a main in bin/ppp'

__END__

=pod

=encoding UTF-8

=head1 NAME

App::pandoc::preprocess - Preprocess Pandoc before Processing Pandoc

=head1 VERSION

version 0.9.5

=head1 ppp - pandoc pre-process

=head1 USAGE

     cat chapters/input-*.pandoc | ppp | pandoc -o output.pdf --smart [more pandoc options...]

Additionally see `etcE<sol>input.txt` for concrete examples.

=head1 PREREQUISITES

=over

=item *

dotE<sol>neato (neato is new!)

=item *

rdfdot

=item *

ditaa

=item *

Image::Magick (for downscaling of large images)

=back

=head1 BACKGROUND

=over

=item *

much simpler design than version 1: pipeable & chainable, reading line-by-line

=item *

parallelized work on image file creation

=back

=head2 How it works

1. while-loop will iterate line by line and is using the flip-flop-operator:

     * as soon, as a ditaa/rdfdot/dot-block starts,
       globals ($fileno, $outfile, etc) are set, so all other routines can see them
     * when actually *inside* the block, the block's contents are printed
       to the newly generated file (image-X.(ditaa/rdfdot/dot))

2. once the flip-flop-operator hits the end of the ditaaE<sol>rdfdotE<sol>dot-block,
a child will be spawned to take over the actual ditaaE<sol>rdfdotE<sol>dot-process
to create the png-file and the globals are reset

3. all other lines which are not part of a ditaaE<sol>rdfdotE<sol>dot-block will simply
be piped through to stdout

4. at the end of the program, all children are waited for

5. in the meantime, the new pandoc contents are printed to stdout

6. all child-processes will remain quiert as far as stdout is concerned and
write to their individual log-files

=head2 Todo

=over

=item *

Captions

=item *

Checks whether ditaa... are available

=item *

check whether ditaa has file.encoding set

=item *

bundle ditaa with this

=back

=head1 AUTHOR

DBR <dbr@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by DBR.

This is free software, licensed under:

  DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE, Version 2, December 2004

=cut


lib/App/pandoc/preprocess/File.pm
package App::pandoc::preprocess::File;
BEGIN {
  $App::pandoc::preprocess::File::AUTHORITY = 'cpan:DBR';
}
{
  $App::pandoc::preprocess::File::VERSION = '0.3.1';
}

#  PODNAME: App::pandoc::preprocess::File
# ABSTRACT: Internal class representing a file

use v5.14;
use strict;
use warnings;

use Moo;
use MooX::Options;
use MooX::Types::MooseLike::Base qw(:all);
use namespace::clean;
use Path::Class;

use Data::Printer;

state $file_count = 0;

sub BUILDARGS {
  my ( $class, @args ) = @_;

  die "Wrong arguments: $!" if @args % 2 == 1;

  return {
    file_count => ++$file_count,
    @args
  };
};

# supported output formats:
# dot: ps svg svgz fig mif hpgl pcl png gif dia imap cmapx
# ditaa: png only?
# rdfdot: --same as dot?-- compare: `rdfdot -output`
sub BUILD {
    my $self = shift;
    my $dir = dir($self->output_directory);
    $dir->mkpath unless -e $dir;
    $self->current_file();
    $self->build_image_file();
    $self->do_downscale_image() if $self->downscale_image();
}

has file_count => (is => 'ro', isa => Int); # default => state $file
has current_image => (is => 'lazy', isa => Str); # e.g. image-x.ditaa
has current_file => (is => 'lazy', isa => Str); # .e.g. is where ditaa/dot is printed to
has output_filename => (is => 'lazy', isa => Str ); # e.g. image-x.png
has downscale_image => (is => 'rw', isa => Bool, default => 0);
has current_format => (is => 'rw', isa => Str);
has content => (is => 'rw', isa => Str);
has output_directory => (
  is => 'rw',
  isa => Str,
  default => '.',
);
has image_generator => (
  is => 'rw',
  isa => HashRef[CodeRef],
  default => sub { +{
      dot   => sub {
        my $self = shift;
        my $cmd = "dot -Tpng -Gsize=1,2\\! -o @{[$self->output_filename]} @{[$self->current_image]}";
        qx/$cmd/;
      },
      ditaa => sub {
        my $self = shift;
        my $cmd = "ditaa -e UTF-8 @{[$self->current_image]} @{[$self->output_filename]}"; #`; #--no-shadows --scale 0.4
        qx/$cmd/;
      },
      rdfdot => sub {
        my $self = shift;
        my $cmd = "rdfdot -ttl @{[$self->current_image]} @{[$self->output_filename]}"; #--no-shadows --scale 0.4
        qx/$cmd/;
      }
    }
  }
);

sub _build_current_image {
  my $self = shift;
  file($self->output_directory, "image-".$self->file_count.".".$self->current_format)->stringify
}

sub _build_output_filename {
  my $self = shift;
  file($self->output_directory, "image-".$self->file_count."."."png")->stringify
}

sub _build_current_file {
  my $self = shift;
  open my $outfile, '>', $self->current_image or die $!;
  print {$outfile} $self->content;
}

sub build_image_file {
  my $self = shift;
  $self->image_generator->{$self->current_format}->($self); # Hack: we need to pass in $self here
}

sub current_image_include {
  my $self = shift;
  return "![](@{[$self->output_filename]})" #\\
}

sub do_downscale_image {
  my $self = shift;
  `mogrify -scale 75% @{[$self->output_filename]}`;
}

1;

__END__

=pod

=head1 NAME

App::pandoc::preprocess::File - Internal class representing a file

=head1 VERSION

version 0.3.1

=head1 AUTHOR

DBR <dbr@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by DBR.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/App/pandoc/preprocess/Checks.pm
package App::pandoc::preprocess;
BEGIN {
  $App::pandoc::preprocess::AUTHORITY = 'cpan:DBR';
}
{
  $App::pandoc::preprocess::VERSION = '0.3.1';
}

#  PODNAME: App::pandoc::preprocess::Checks
# ABSTRACT: Checks your environment

use v5.14;
use strict;
use warnings;

use Moo;
use MooX::Types::MooseLike::Base qw| :all |;

has installed_ditaa   => (is => 'ro', isa => Defined, required => 1, default => sub { `which ditaa`   or die q<Can't find program ditaa -- Abort>  });
has installed_rdfdot  => (is => 'ro', isa => Defined, required => 1, default => sub { `which rdfdot`  or die q<Can't find program rdfdot -- Abort> });
has installed_dot     => (is => 'ro', isa => Defined, required => 1, default => sub { `which dot`     or die q<Can't find program dot -- Abort>    });
has installed_mogrify => (is => 'ro', isa => Defined, required => 1, default => sub { `which mogrify` or die q<Can't find program mogrify -- Abort>});

1;

__END__

=pod

=head1 NAME

App::pandoc::preprocess::Checks - Checks your environment

=head1 VERSION

version 0.3.1

=head1 AUTHOR

DBR <dbr@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by DBR.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Preprocess Pandoc before Processing Pandoc",
  "AUTHOR" => "DBR <dbr\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "App-pandoc-preprocess",
  "EXE_FILES" => [
    "bin/ppp",
    "bin/ppp"
  ],
  "LICENSE" => "unrestricted",
  "NAME" => "App::pandoc::preprocess",
  "PREREQ_PM" => {
    "File::Path" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Scalar::Util" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.94",
    "Test::UseAllModules" => 0
  },
  "VERSION" => "0.9.5",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Path" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "Module::Build" => "0.3601",
  "Scalar::Util" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.94",
  "Test::UseAllModules" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



